src/components/vm/consoles/utils.js

```diff
@@ -0,0 +1,43 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2023 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+export function getServerAddress() {
+    return window.location.hostname;
```

Comment:
@martinpitt Is there any way for me to mock window.location.hostname so I can test a situation where it's value is different from hostname/127.x.x.x ?

Reply:
Not directly -- if you write to that attribute, you will actually change the page. But you can add a destructive test with `provision` to *actually* get a second machine, like you already did in test/check-machines-migrate. That's IMHO a better way to validate that this really works.

Reply:
As you just said in the chat, this won't help, as the browser runs outside the VMs and the cockpit port *always* gets tunneled to 127.0.0.2. We can't run a browser inside the VM. So this is actually exactly the unfixable [cockpit behind a reverse proxy](https://github.com/cockpit-project/cockpit-machines/pull/1084#discussion_r1205002727) scenario below.

But there's a workaround: You can restrict the "localhost" check to 127.0.0.1, treat any other 127.* address as "remote", and then have one test set up a `socat` from 127.0.0.2:cockpit_port to 127.0.0.1:5555 (or any other port which is definitively not used); then that test can connect the browser to that forwarded port and it'll appear as truly "localhost". See e.g. [this cockpit test](https://github.com/cockpit-project/cockpit/blob/main/test/verify/check-connection#L940) where we do this. The caveat is that we run socat in the VM in these tests. But (1) our [tasks container has socat](https://github.com/cockpit-project/cockpituous/blob/main/tasks/Dockerfile#L52), and (2) you can also use `ssh -L` to tunnel the port (which is incidentally the very thing that this PR advertises, so perhaps that's better :grin: )